src/components/engagement/handlers/set-initial-end-date.handler.ts

```diff
@@ -0,0 +1,85 @@
+import { ServerException } from '../../../common';
+import {
+  DatabaseService,
+  EventsHandler,
+  IEventHandler,
+  ILogger,
+  Logger,
+} from '../../../core';
+import { EngagementStatus } from '../dto';
+import { EngagementService } from '../engagement.service';
+import {
+  InternshipEngagementCreatedEvent,
+  InternshipEngagementUpdatedEvent,
+  LanguageEngagementCreatedEvent,
+  LanguageEngagementUpdatedEvent,
+} from '../events';
+
+type SubscribedEvent =
+  | LanguageEngagementCreatedEvent
+  | LanguageEngagementUpdatedEvent
+  | InternshipEngagementCreatedEvent
+  | InternshipEngagementUpdatedEvent;
+
+@EventsHandler(
+  LanguageEngagementCreatedEvent,
+  LanguageEngagementUpdatedEvent,
+  InternshipEngagementCreatedEvent,
+  InternshipEngagementUpdatedEvent
+)
+export class SetInitialEndDate implements IEventHandler<SubscribedEvent> {
+  constructor(
+    private readonly db: DatabaseService,
+    private readonly engagementService: EngagementService,
+    @Logger('engagement:set-initial-end-date') private readonly logger: ILogger
+  ) {}
+
+  async handle(event: SubscribedEvent) {
+    this.logger.debug('Engagement mutation, set initial end date', {
+      ...event,
+      event: event.constructor.name,
+    });
+
+    try {
+      const shouldUpdateInitialEndDate =
+        event.engagement.status === EngagementStatus.Active &&
+        event.engagement.initialEndDate.value == null &&
+        event.engagement.endDate.value != null;
+
+      if (shouldUpdateInitialEndDate) {
+        const updateInput = {
+          id: event.engagement.id,
+          initialEndDate: event.engagement.endDate.value!,
+        };
+
+        if (
+          event instanceof LanguageEngagementCreatedEvent ||
+          event instanceof LanguageEngagementUpdatedEvent
+        ) {
+          await this.engagementService.updateLanguageEngagement(
+            updateInput,
+            event.session
+          );
```

Comment:
By setting this here the engagement object returned from the API will be stale. Since the `initialEndDate` will have the previous value.
Maybe the event object could have a needsRefresh flag so that the engagement service can check for that and refetch it again.
Or maybe the event could have a setter for the engagement object and then the service returns that? That would eliminate an extra db call.

Reply:
@CarsonF I updated so that after calling the event handler it will always fetch a fresh object. This will result in extra db reads, however the alternatives I could see all required rewriting the event bus so that the handler could set a flag to be returned from the publish function. 

The current interface returns void so I couldn't see a way to communicate out of the handler to the engagement service.
```
  async publish<T extends IEvent>(event: T): Promise<void> {
    ...
  }
```

I didn't want to update the event bus since that would effect lots of code outside of my narrow focus on this PR. Carson, let me know if you want me to focus on optimizing this more for this PR, or also let me know if I missed a trick here and there is a simpler path then the one I went down.

Reply:
the publish returns void but the event can still have state.

```tsx
const event = new EngagementUpdated(...);
this.eventBus.publish(event);
if (event.isRefreshNeeded) {
  // do another read
}

// in handler
event.needsRefresh();
```

Or instead of requiring a blind refresh we could have the `updated` object be settable and that used for the output.
```tsx
// service
const event = ...;
publish(event);
return event.updated;

// handler
event.updated = {...};
```

Reply:
Ah, I understand now. I'll make that update.